XenAdmin/SettingsPanels/SnmpEditPage.cs

```diff
@@ -0,0 +1,278 @@
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR 
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, 
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, 
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR 
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, 
+ * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING 
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE 
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF 
+ * SUCH DAMAGE.
+ */
+
+using System;
+using System.Drawing;
+using System.Windows.Forms;
+using XenAPI;
+using XenAdmin.Actions;
+using XenAdmin.Core;
+using System.Text.RegularExpressions;
+using XenAdmin.Actions.SNMP;
+
+namespace XenAdmin.SettingsPanels
+{
+    public partial class SnmpEditPage : UserControl, IEditPage
+    {
+        public string SubText => Messages.SNMP_EDIT_PAGE_TEXT;
+        public Image Image => Images.StaticImages._000_Network_h32bit_16;
+        private ToolTip _invalidParamToolTip = new ToolTip
+        {
+            IsBalloon = true,
+            ToolTipIcon = ToolTipIcon.Warning,
+        };
+        private string _invalidParamToolTipText = "";
+        private static readonly Regex RegexCommon = new Regex(@"^[a-zA-Z0-9-.\#@=:_]{6,32}$");
+        private static readonly Regex RegexEncryptTextBox = new Regex(@"^([a-zA-Z0-9-.\#@=:_]{8,32}|[*]{8})$");
+        private static bool _encryptTextBoxFlag = false;
+        private IXenObject _clone;
+        private readonly SnmpConfiguration _snmpConfiguration = new SnmpConfiguration();
+        private readonly SnmpConfiguration _snmpCurrentConfiguration = new SnmpConfiguration();
+        public bool HasChanged
+        {
+            get
+            {
+                UpdateCurrentSnmpConfiguration();
+                return !_snmpCurrentConfiguration.Equals(_snmpConfiguration);
+            }
+        }
+        public bool ValidToSave
+        {
+            get
+            {
+                _invalidParamToolTip.Dispose();
+                _invalidParamToolTip = new ToolTip
+                {
+                    IsBalloon = true,
+                    ToolTipIcon = ToolTipIcon.Warning,
+                };
+                _invalidParamToolTipText = " ";
+
+                if (!EnableSnmpCheckBox.Checked)
+                {
+                    return true;
+                }
+                if (!SupportV2cCheckBox.Checked && !SupportV3CheckBox.Checked)
+                {
+                    _invalidParamToolTip.Tag = SupportV2cCheckBox;
+                    _invalidParamToolTip.ToolTipTitle = Messages.SNMP_ALLOW_CHOOSE_TITLE;
+                    return false;
+                }
+                if (SupportV2cCheckBox.Checked)
+                {
+                    var communityStr = CommunityTextBox.Text;
+                    if (!RegexCommon.Match(communityStr.Trim()).Success)

```

Comment:
Shouldn't we be checking for null first? I.e. `if (string.IsNullOrEmpty(communityStr) || !Regex...)`. Same in the other checks below.